src/server.rs

```diff
@@ -368,57 +341,57 @@ async fn index(
     body: Bytes,
     req: HttpRequest,
 ) -> impl Responder {
-    let mut request = Request::from_actix_request(&req, body);
-
-    apply_middleware(
-        &mut request,
-        MiddlewareRoute::BeforeRequest,
-        &middleware_router,
-        req.uri().path(),
-    )
-    .await;
-
-    let mut response_builder = HttpResponse::Ok();
-    apply_dashmap_headers(&mut response_builder, &global_request_headers);
-    apply_dashmap_headers(&mut response_builder, &global_response_headers);
-    apply_hashmap_headers(&mut response_builder, &request.headers);
-
-    let response = if let Some(r) = const_router.get_route(req.method(), req.uri().path()) {
-        apply_hashmap_headers(&mut response_builder, &r.headers);
-        response_builder
-            .status(StatusCode::from_u16(r.status_code).unwrap())
-            .body(r.body)
+    let mut request = Request::from_actix_request(&req, body, &global_request_headers);
+
+    // Before middleware
+    request = match middleware_router.get_route(&MiddlewareRoute::BeforeRequest, req.uri().path()) {
+        Some((function, route_params)) => {
+            request.params = route_params;
+            execute_middleware_function(&request, function)
+                .await
+                .unwrap_or_else(|e| {
+                    error!("Error while executing before middleware: {e}");
+                    request
+                })
+        }
+        None => request,
+    };
+
+    // Route execution
+    let mut response = if let Some(r) = const_router.get_route(req.method(), req.uri().path()) {
```

Comment:
one small suggestion

can we rename `r` to `res` here? It will make slightly better. This is a personal quirk